cmd/sourced/cmd/web.go

```diff
@@ -32,54 +33,177 @@ func init() {
-	var stdout bytes.Buffer
-	// wait for the container to start, it can take a while in some cases
-	for {
-		err := compose.RunWithIO(context.Background(),
-			os.Stdin, &stdout, nil, "port", containerName, "8088")
+func openUI(address string) error {
+	// docker-compose returns 0.0.0.0 which is correct for the bind address
+	// but incorrect as connect address
+	url := fmt.Sprintf("http://%s", strings.Replace(address, "0.0.0.0", "127.0.0.1", 1))
 
-		if err == nil {
+	for {
+		client := http.Client{Timeout: time.Second}
+		if _, err := client.Get(url); err == nil {
 			break
 		}
 
-		if workdir.ErrMalformed.Is(err) || dir.ErrNotExist.Is(err) || dir.ErrNotValid.Is(err) {
-			return err
-		}
-
 		time.Sleep(1 * time.Second)
 	}
 
-	address := strings.TrimSpace(stdout.String())
-	if address == "" {
-		return fmt.Errorf("could not find the public port of %s", containerName)
+	if err := browser.OpenURL(url); err != nil {
+		return errors.Wrap(err, "could not open the browser")
 	}
 
-	// docker-compose returns 0.0.0.0 which is correct for the bind address
-	// but incorrect as connect address
-	url := fmt.Sprintf("http://%s", strings.Replace(address, "0.0.0.0", "127.0.0.1", 1))
+	return nil
+}
+
+func checkFailFast(stdout *bytes.Buffer) (bool, error) {
+	err := compose.RunWithIO(context.Background(),
+		os.Stdin, stdout, nil, "port", containerName, "8088")
+	if workdir.ErrMalformed.Is(err) || dir.ErrNotExist.Is(err) || dir.ErrNotValid.Is(err) {
+		return true, err
+	}
+
+	if err != nil {
+		return false, err
+	}
+
+	return false, nil
+}
 
+func waitForContainer(stdout *bytes.Buffer) {
 	for {
-		client := http.Client{Timeout: time.Second}
-		if _, err := client.Get(url); err == nil {
+		if err := compose.RunWithIO(context.Background(),
+			os.Stdin, stdout, nil, "port", containerName, "8088"); err == nil {
 			break
 		}
 
 		time.Sleep(1 * time.Second)
 	}
+}
 
-	if err := browser.OpenURL(url); err != nil {
-		return errors.Wrap(err, "could not open the browser")
+var newLineFormatter = regexp.MustCompile(`(\r\n|\r|\n)`)
+var stateExtractor = regexp.MustCompile(`(?m)^srcd-[\w\d]+.*(Up|Exit (\w+))`)
+
+func normalizeNewLine(s string) string {
+	return newLineFormatter.ReplaceAllString(s, "\n")
+}
+
+// runMonitor checks the status of the containers in order to early exit in case
+// an unrecoverable error occurs.
+// The monitoring is performed by running `docker-compose ps <service>` for each
+// service returned by `docker-compose config --services`, and by grepping the
+// state from the stdout using a regex.
+// Getting the state of all the containers in a single pass by running `docker-compose ps`
+// and by using a multi-line regex to extract both service name and state is not reliable.
+// The reason is that the prefix of a container can be very long, especially for local
+// initialization, due to the value that we set for `COMPOSE_PROJECT_NAME` env var, and
+// docker-compose may split the name into multiple lines.
+// E.g.:
```

Comment:
Do you think that the example is too much?

Reply:
yes :) just 1 line is enough to understand.

Reply:
Done [here](https://github.com/src-d/sourced-ce/pull/186/commits/f98097944933f458fe2ceed62ce80173cd6b2c2c).